File: game-server/trading.py
"""Trading system for Gradient Bang.

Handles commodity pricing, port inventory management, and trade transactions.
"""

import math
from typing import Dict, Optional, Tuple
from enum import Enum

# Commodity base prices
BASE_PRICES = {
    "fuel_ore": 25,
    "organics": 10,
    "equipment": 40
}

# Price bands for port trading
SELL_MIN = 0.75  # Port sells to player at 75% when full stock
SELL_MAX = 1.10  # Port sells to player at 110% when low stock
BUY_MIN = 0.90   # Port buys from player at 90% when low demand
BUY_MAX = 1.30   # Port buys from player at 130% when high demand

# Port inventory defaults (matching universe-bang.py)
PORT_DEFAULT_CAP = 1000
PORT_STARTING_FILL = 0.70

# Regeneration rates for port inventory
REGEN_FRACTION_STOCK = 0.25   # 25% of max per regeneration
REGEN_FRACTION_DEMAND = 0.25  # 25% of max per regeneration


class TradeType(Enum):
    """Type of trade transaction."""
    BUY = "buy"    # Player buys from port
    SELL = "sell"  # Player sells to port


class TradingError(Exception):
    """Base exception for trading errors."""
    pass


def calculate_price_sell_to_player(commodity: str, stock: int, max_capacity: int) -> int:
    """Calculate the price a port sells to a player.
    
    Uses sqrt curve: price rises slowly when stock is high, spikes when low.
    
    Args:
        commodity: Commodity type (fuel_ore, organics, equipment)
        stock: Current stock level
        max_capacity: Maximum stock capacity
        
    Returns:
        Price in credits
        
    Raises:
        ValueError: If commodity is unknown or max_capacity is 0
    """
    if commodity not in BASE_PRICES:
        raise ValueError(f"Unknown commodity: {commodity}")
    if max_capacity <= 0:
        raise ValueError("Max capacity must be positive")
    
    # Calculate scarcity (0 = full stock, 1 = empty)
    fullness = stock / max_capacity
    scarcity = 1 - fullness
    
    # Use sqrt curve for more realistic pricing
    # Prices rise slowly at first, then spike near depletion
    price_multiplier = SELL_MIN + (SELL_MAX - SELL_MIN) * math.sqrt(scarcity)
    
    return int(round(BASE_PRICES[commodity] * price_multiplier))


def calculate_price_buy_from_player(commodity: str, stock: int, max_capacity: int) -> int:
    """Calculate the price a port buys from a player.
    
    Uses sqrt curve: price is high when stock is low (need more), low when stock is high (saturated).
    
    Args:
        commodity: Commodity type (fuel_ore, organics, equipment)
        stock: Current stock level
        max_capacity: Maximum stock capacity
        
    Returns:
        Price in credits
        
    Raises:
        ValueError: If commodity is unknown or max_capacity is 0
    """
    if commodity not in BASE_PRICES:
        raise ValueError(f"Unknown commodity: {commodity}")
    if max_capacity <= 0:
        raise ValueError("Max capacity must be positive")
    
    # Calculate need (0 = full/saturated, 1 = empty/desperate)
    fullness = stock / max_capacity
    need = 1 - fullness  # Lower stock = higher need = higher price
    
    # Use sqrt curve for more realistic pricing
    # Prices drop slowly at first, then plunge near saturation
    price_multiplier = BUY_MIN + (BUY_MAX - BUY_MIN) * math.sqrt(need)
    
    return int(round(BASE_PRICES[commodity] * price_multiplier))


def get_port_prices(port_data: Dict) -> Dict[str, Optional[int]]:
    """Get current prices for all commodities at a port.
    
    Args:
        port_data: Port data dictionary with stock/max_capacity info
        
    Returns:
        Dictionary mapping commodities to their prices (None if not traded).
        The price is for selling TO player if port sells it, or buying FROM player if port buys it.
    """
    # Map commodity names to their keys in port data
    commodity_keys = {
        "fuel_ore": "FO",
        "organics": "OG",
        "equipment": "EQ"
    }
    
    prices = {}
    
    for commodity, key in commodity_keys.items():
        price = None
        
        stock = port_data.get("stock", {}).get(key, 0)
        max_capacity = port_data.get("max_capacity", {}).get(key, 0)
        
        # Check if port sells this commodity (price is what player pays)
        if commodity in port_data.get("sells", []):
            if max_capacity > 0:
                price = calculate_price_sell_to_player(commodity, stock, max_capacity)
        
        # Check if port buys this commodity (price is what player receives)
        elif commodity in port_data.get("buys", []):
            # Port can only buy if it has room (stock < max_capacity)
            if max_capacity > 0 and stock < max_capacity:
                price = calculate_price_buy_from_player(commodity, stock, max_capacity)
        
        prices[commodity] = price
    
    return prices


def validate_buy_transaction(
    player_credits: int,
    cargo_used: int,
    cargo_capacity: int,
    commodity: str,
    quantity: int,
    port_stock: int,
    price_per_unit: int
) -> None:
    """Validate a buy transaction can be completed.
    
    Args:
        player_credits: Player's current credits
        cargo_used: Current cargo space used
        cargo_capacity: Maximum cargo capacity
        commodity: Commodity to buy
        quantity: Quantity to buy
        port_stock: Port's current stock
        price_per_unit: Price per unit
        
    Raises:
        TradingError: If transaction cannot be completed
    """
    if quantity <= 0:
        raise TradingError("Quantity must be positive")
    
    if port_stock < quantity:
        raise TradingError(f"Port only has {port_stock} units of {commodity}")
    
    free_space = cargo_capacity - cargo_used
    if free_space < quantity:
        raise TradingError(f"Not enough cargo space. Available: {free_space}")
    
    total_cost = price_per_unit * quantity
    if player_credits < total_cost:
        raise TradingError(f"Insufficient credits. Need {total_cost}, have {player_credits}")


def validate_sell_transaction(
    player_cargo: Dict[str, int],
    commodity: str,
    quantity: int,
    port_stock: int,
    port_max_capacity: int
) -> None:
    """Validate a sell transaction can be completed.
    
    Args:
        player_cargo: Player's current cargo
        commodity: Commodity to sell
        quantity: Quantity to sell
        port_stock: Port's current stock
        port_max_capacity: Port's maximum capacity
        
    Raises:
        TradingError: If transaction cannot be completed
    """
    if quantity <= 0:
        raise TradingError("Quantity must be positive")
    
    if player_cargo.get(commodity, 0) < quantity:
        raise TradingError(f"Not enough {commodity} to sell. Have {player_cargo.get(commodity, 0)}")
    
    available_capacity = port_max_capacity - port_stock
    if available_capacity < quantity:
        raise TradingError(f"Port can only buy {available_capacity} units of {commodity}")

File: deployment/supabase/functions/_shared/trading.ts
import type { SupabaseClient } from "@supabase/supabase-js";

export type Commodity = "quantum_foam" | "retro_organics" | "neuro_symbolics";
export type CommodityKey = "QF" | "RO" | "NS";

const COMMODITY_ORDER: Commodity[] = [
  "quantum_foam",
  "retro_organics",
  "neuro_symbolics",
];
const COMMODITY_KEY_MAP: Record<Commodity, CommodityKey> = {
  quantum_foam: "QF",
  retro_organics: "RO",
  neuro_symbolics: "NS",
};

const BASE_PRICES: Record<Commodity, number> = {
  quantum_foam: 25,
  retro_organics: 10,
  neuro_symbolics: 40,
};

const SELL_MIN = 0.75;
const SELL_MAX = 1.1;
const BUY_MIN = 0.9;
const BUY_MAX = 1.3;

export interface PortRow {
  port_id: number;
  sector_id: number;
  port_code: string;
  port_class: number;
  max_qf: number;
  max_ro: number;
  max_ns: number;
  stock_qf: number;
  stock_ro: number;
  stock_ns: number;
  version: number;
  last_updated: string | null;
}

export interface PortData {
  code: string;
  class: number;
  stock: Record<CommodityKey, number>;
  max_capacity: Record<CommodityKey, number>;
  buys: Commodity[];
  sells: Commodity[];
}

export type TradeType = "buy" | "sell";

export function isCommodity(candidate: unknown): candidate is Commodity {
  return (
    typeof candidate === "string" &&
    (COMMODITY_ORDER as string[]).includes(candidate)
  );
}

export function normalizeCommodity(candidate: unknown): Commodity | null {
  if (isCommodity(candidate)) {
    return candidate;
  }
  if (typeof candidate === "string") {
    const lowered = candidate.toLowerCase();
    return isCommodity(lowered) ? (lowered as Commodity) : null;
  }
  return null;
}

export function commodityKey(value: Commodity): CommodityKey {
  return COMMODITY_KEY_MAP[value];
}

export function commodityFromIndex(index: number): Commodity {
  return COMMODITY_ORDER[index] ?? "quantum_foam";
}

export async function loadPortBySector(
  supabase: SupabaseClient,
  sectorId: number,
): Promise<PortRow | null> {
  const { data: contents, error: contentsError } = await supabase
    .from("sector_contents")
    .select("port_id")
    .eq("sector_id", sectorId)
    .maybeSingle();
  if (contentsError) {
    throw new Error(
      `failed to load sector contents for ${sectorId}: ${contentsError.message}`,
    );
  }
  if (!contents?.port_id) {
    return null;
  }

  const { data, error } = await supabase
    .from("ports")
    .select(
      "port_id, sector_id, port_code, port_class, max_qf, max_ro, max_ns, stock_qf, stock_ro, stock_ns, version, last_updated",
    )
    .eq("port_id", contents.port_id)
    .maybeSingle();
  if (error) {
    throw new Error(
      `failed to load port for sector ${sectorId}: ${error.message}`,
    );
  }
  return data as PortRow | null;
}

export function buildPortData(row: PortRow): PortData {
  const buys: Commodity[] = [];
  const sells: Commodity[] = [];
  for (let i = 0; i < COMMODITY_ORDER.length; i += 1) {
    const commodity = commodityFromIndex(i);
    const code = row.port_code?.charAt(i) ?? "S";
    if (code === "B") {
      buys.push(commodity);
    } else {
      sells.push(commodity);
    }
  }
  return {
    code: row.port_code,
    class: row.port_class,
    stock: {
      QF: row.stock_qf,
      RO: row.stock_ro,
      NS: row.stock_ns,
    },
    max_capacity: {
      QF: row.max_qf,
      RO: row.max_ro,
      NS: row.max_ns,
    },
    buys,
    sells,
  };
}

export function portSupportsTrade(
  portData: PortData,
  commodity: Commodity,
  tradeType: TradeType,
): boolean {
  if (tradeType === "buy") {
    return portData.sells.includes(commodity);
  }
  return portData.buys.includes(commodity);
}

export function calculatePriceSellToPlayer(
  commodity: Commodity,
  stock: number,
  maxCapacity: number,
): number {
  if (maxCapacity <= 0) {
    throw new Error("invalid max capacity for sell price calculation");
  }
  const fullness = stock / maxCapacity;
  const scarcity = 1 - fullness;
  const multiplier =
    SELL_MIN +
    (SELL_MAX - SELL_MIN) * Math.sqrt(Math.max(0, Math.min(1, scarcity)));
  return Math.round(BASE_PRICES[commodity] * multiplier);
}

export function calculatePriceBuyFromPlayer(
  commodity: Commodity,
  stock: number,
  maxCapacity: number,
): number {
  if (maxCapacity <= 0) {
    throw new Error("invalid max capacity for buy price calculation");
  }
  const need = 1 - stock / maxCapacity;
  const multiplier =
    BUY_MIN + (BUY_MAX - BUY_MIN) * Math.sqrt(Math.max(0, Math.min(1, need)));
  return Math.round(BASE_PRICES[commodity] * multiplier);
}

export function getPortPrices(
  portData: PortData,
): Record<Commodity, number | null> {
  const prices: Record<Commodity, number | null> = {
    quantum_foam: null,
    retro_organics: null,
    neuro_symbolics: null,
  };
  for (const commodity of COMMODITY_ORDER) {
    const key = commodityKey(commodity);
    const stock = portData.stock[key];
    const maxCapacity = portData.max_capacity[key];
    if (portData.sells.includes(commodity)) {
      prices[commodity] = calculatePriceSellToPlayer(
        commodity,
        stock,
        maxCapacity,
      );
    } else if (portData.buys.includes(commodity) && stock < maxCapacity) {
      prices[commodity] = calculatePriceBuyFromPlayer(
        commodity,
        stock,
        maxCapacity,
      );
    } else {
      prices[commodity] = null;
    }
  }
  return prices;
}

export function getPortStock(portData: PortData): Record<Commodity, number> {
  return {
    quantum_foam: portData.stock.QF ?? 0,
    retro_organics: portData.stock.RO ?? 0,
    neuro_symbolics: portData.stock.NS ?? 0,
  };
}

export class TradingValidationError extends Error {
  status: number;

  constructor(message: string, status = 400) {
    super(message);
    this.name = "TradingValidationError";
    this.status = status;
  }
}

export function validateBuyTransaction(
  playerCredits: number,
  cargoUsed: number,
  cargoCapacity: number,
  commodity: Commodity,
  quantity: number,
  portStock: number,
  pricePerUnit: number,
): void {
  if (!Number.isInteger(quantity) || quantity <= 0) {
    throw new TradingValidationError("Quantity must be a positive integer");
  }
  if (portStock < quantity) {
    throw new TradingValidationError(
      `Port only has ${portStock} units of ${commodity}`,
    );
  }
  const freeSpace = cargoCapacity - cargoUsed;
  if (freeSpace < quantity) {
    throw new TradingValidationError(
      `Not enough cargo space. Available: ${freeSpace}`,
    );
  }
  const totalPrice = pricePerUnit * quantity;
  if (playerCredits < totalPrice) {
    throw new TradingValidationError(
      `Insufficient credits. Need ${totalPrice}, have ${playerCredits}`,
    );
  }
}

export function validateSellTransaction(
  cargo: Record<Commodity, number>,
  commodity: Commodity,
  quantity: number,
  portStock: number,
  portMaxCapacity: number,
): void {
  if (!Number.isInteger(quantity) || quantity <= 0) {
    throw new TradingValidationError("Quantity must be a positive integer");
  }
  if ((cargo[commodity] ?? 0) < quantity) {
    throw new TradingValidationError(
      `Not enough ${commodity} to sell. Have ${cargo[commodity] ?? 0}`,
    );
  }
  const availableCapacity = portMaxCapacity - portStock;
  if (availableCapacity < quantity) {
    throw new TradingValidationError(
      `Port can only buy ${availableCapacity} units of ${commodity}`,
    );
  }
}
